File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.cpp
class Solution {
public:
    int dir[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int n = heights.size();
        int m = heights[0].size();
        vector<vector<bool>> pacific(n,vector<bool>(m));
        vector<vector<bool>> atlantic(n,vector<bool>(m));

        for(int i=0;i<n;i++){
            dfs(i,0,pacific,heights);
            dfs(i,m-1,atlantic,heights);
        }
        for(int i=0;i<m;i++){
            dfs(0,i,pacific,heights);
            dfs(n-1,i,atlantic,heights);
        }
        vector<vector<int>> res;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(atlantic[i][j] && pacific[i][j])res.push_back({i,j});
            }
        }
        return res;


    }
    void dfs(int r, int c, vector<vector<bool>>& ocean, vector<vector<int>>& heights){
        ocean[r][c]=1;
        for(int i=0;i<4;i++){
            int nr = r+dir[i][0];
            int nc = c+dir[i][1];
            if(nr>=0 && nr<heights.size() && nc>=0 && nc<heights[0].size() && !ocean[nr][nc] && heights[nr][nc]>=heights[r][c]){
                dfs(nr,nc,ocean,heights);
            }
        }
    }
};

File: Data Structures & Algorithms/search-for-word/submission-0.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        n = len(board)
        m = len(board[0])

        def dfs(r,c,i):
            if i==len(word):
                return True
            if (r<0 or c<0 or r==n or c==m or word[i]!=board[r][c] or board[r][c]=="#"):
                return False
            board[r][c]="#"

            res = dfs(r+1,c,i+1) or dfs(r,c+1,i+1) or dfs(r-1,c,i+1) or dfs(r,c-1,i+1)

            board[r][c]=word[i]
            return res

        for i in range(n):
            for j in range(m):
                if dfs(i,j,0):
                    return True

        return False        
